src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3RequestTests.cs

```diff
@@ -108,6 +109,88 @@ public async Task POST_ServerCompletsWithoutReadingRequestBody_ClientGetsRespons
+        public async Task POST_ClientCancellationUpload_RequestAbortRaised()
+        {
+            // Arrange
+            var syncPoint = new SyncPoint();
+            var cancelledTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+            var readAsyncTask = new TaskCompletionSource<Task>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            var builder = CreateHttp3HostBuilder(async context =>
+            {
+                context.RequestAborted.Register(() => cancelledTcs.SetResult());
+
+                var body = context.Request.Body;
+
+                // Read content
+                var data = new List<byte>();
+                var buffer = new byte[1024];
+                var readCount = 0;
+                while ((readCount = await body.ReadAsync(buffer).DefaultTimeout()) != -1)
+                {
+                    data.AddRange(buffer.AsMemory(0, readCount).ToArray());
+                    if (data.Count == TestData.Length)
+                    {
+                        break;
+                    }
+                }
+
+                // Sync with client
+                await syncPoint.WaitToContinue();
+
+                // Wait for task cancellation
+                await cancelledTcs.Task;
+
+                readAsyncTask.SetResult(body.ReadAsync(buffer).AsTask());
+            });
+
+            using (var host = builder.Build())
+            using (var client = new HttpClient())
+            {
+                await host.StartAsync();
+
+                var cts = new CancellationTokenSource();
+                var requestContent = new StreamingHttpContext();
+
+                var request = new HttpRequestMessage(HttpMethod.Post, $"https://127.0.0.1:{host.GetPort()}/");
+                request.Content = requestContent;
+                request.Version = HttpVersion.Version30;
+                request.VersionPolicy = HttpVersionPolicy.RequestVersionExact;
+
+                // Act
+                var responseTask = client.SendAsync(request, cts.Token);
+
+                var requestStream = await requestContent.GetStreamAsync();
+
+                // Send headers
+                await requestStream.FlushAsync();
+                // Write content
+                await requestStream.WriteAsync(TestData);
+
+                // Wait until content is read on server
+                await syncPoint.WaitForSyncPoint();
+
+                // Cancel request
+                cts.Cancel();
+
+                // Continue on server
+                syncPoint.Continue();
+
+                // Assert
+                await Assert.ThrowsAnyAsync<OperationCanceledException>(() => responseTask).DefaultTimeout();
+
+                await cancelledTcs.Task.DefaultTimeout();
+
+                var serverWriteTask = await readAsyncTask.Task.DefaultTimeout();
+
+                var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => serverWriteTask).DefaultTimeout();
```

Comment:
If the request is aborted and ReadAsync or WriteAsync are called, what should the exception type be? I think HTTP/2 throws IOException...